src/relay/backend/build_module.cc

```diff
@@ -0,0 +1,669 @@
+    func = _CallPacked("relay._ir_pass.RewriteDeviceAnnotation", func, fallback_idx);
+    auto device_map = _CallPacked<Map<Expr, Integer> >("relay._ir_pass.CollectDeviceInfo",
+                                                       func,
+                                                       nullptr);
+    if (device_map.size() == 0) {
+      auto annotation_map = _CallPacked<Map<Expr, Integer> >("_ir_pass.CollectDeviceAnnotationOps",
+                                                             func,
+                                                             nullptr);
+      if (annotation_map.size() == 0) {
+        targets_map_ptr->Set(
+          ir::IntImm::make(HalideIR::Int(64), 0),
+          ir::StringImm::make(cfg.fall_back_device));
+      } else {
+        int64_t dev_type = -1;
+        for (auto kv : annotation_map) {
+          dev_type = kv.second->value;
+          break;
+        }
+        for (auto kv : annotation_map) {
+          CHECK_EQ(kv.second->value, dev_type)
+            << "Expressions in the function are "
+            << "annotated with various device types,"
+            << "but not device copy operators "
+            << "found. Please check the "
+            << "RewriteAnnotation pass.";
+        }
+        targets_map_ptr->Set(
+          ir::IntImm::make(HalideIR::Int(64), 0),
+          ir::StringImm::make(ContextMap::Mask2Str(dev_type)));
+      }
+    }
+    return func;
+  }
+  /*!
+   * \brief Build module given lowered functions
+   * 
+   * \param lowered_funcs 
+   * \param targets 
+   * \param cfg 
+   */
+  void _BuildModule(Map<std::string, Array<LoweredFunc> > lowered_funcs,
+                    Map<HalideIR::Expr, HalideIR::Expr> targets,
+                    const BuildConfig& cfg) {
+    auto target_host = Target::create(cfg_.fall_back_device);
+    for (const auto& kv : lowered_funcs) {
+      std::unordered_set<std::string> fname_set;
+      for (auto f : kv.second) {
+        if (fname_set.count(f->name)) {
+          LOG(FATAL) << "Duplicate function name "
+                     << f->name;
+        }
+        fname_set.insert(f->name);
+      }
+    }
+    std::unordered_map<std::string, Target> target_map;
+    for (const auto& kv : lowered_funcs) {
+      target_map[kv.first] = Target::create(kv.first);
+    }
+    Array<LoweredFunc> fhost_all;
+    std::vector<runtime::Module> device_module;
+    for (const auto& kv : lowered_funcs) {
+      auto target = target_map[kv.first];
+      auto mdev = build(kv.second,
+                        target,
+                        target_host,
+                        cfg,
+                        &fhost_all,
+                        &device_module);
+    }
+
+    auto mhost = build(fhost_all,
+                       target_host,
+                       target_host,
+                       cfg);
+
+    for (auto mdev : device_module) {
+      mhost.Import(mdev);
+    }
+    ret_.mod = mhost;
+  }
+
+  /*!
```

Comment:
This comment doesn't exactly match the below signature, can you update it?

Reply:
There's no return of this function. Why do you put return in the doc?

Reply:
Fixed